cobra/flux_analysis/gapfilling.py

```diff
@@ -1,9 +1,176 @@
 # -*- coding: utf-8 -*-
 
-from __future__ import absolute_import
+from __future__ import absolute_import, print_function
 
-from cobra.exceptions import DefunctError
+from collections import defaultdict
+from sympy import Add
+from warnings import warn
 
+from optlang.interface import OPTIMAL
+from optlang.exceptions import SolverError
+from cobra.core import Model
+from cobra.util import assert_optimal, get_objective_for
 
-def growMatch():
-    raise DefunctError('growMatch', 'modelseed', 'http://modelseed.org/')
+
+class GapFiller(object):
+    """Class for performing gap filling.
+
+    Parameters
+    ----------
+    model : cobra.Model
+        The model to perform gap filling on
+    universal : cobra.Model
+        A universal model with reactions that can be used to complete the model
+    objective : model.interface.Objective
+        The objective that must be feasible in the filled model.
+    lower_bound : float
+        The minimally accepted flux for the objective in the filled model.
+    """
+
+    def __init__(self, model, universal=None, objective=None, lower_bound=0.05,
+                 penalties=None, exchange_reactions=False,
+                 demand_reactions=True):
+        self.original_model = model
+        self.lower_bound = lower_bound
+        self.model = model.copy()
+        self.universal = universal.copy() if universal else Model('universal')
+        self.penalties = dict(universal=1, exchange=100,
+                              demand=1) if not penalties else penalties
+        self.indicators = list()
+        self.costs = defaultdict(int)
+        self.objective = model.objective if objective is None else objective
+        self.extend_model(exchange_reactions, demand_reactions)
+        self.fix_objective_as_constraint(lower_bound)
+        self.add_switches_and_objective()
+
+    def fix_objective_as_constraint(self, lower_bound):
+        constraint = self.model.problem.Constraint(
+            self.objective.expression, lb=lower_bound,
+            name='original_objective_constraint')
+        self.model.add_cons_vars([constraint], sloppy=True)
+
+    def extend_model(self, exchange_reactions=False, demand_reactions=True):
+        for rxn in self.universal.reactions:
+            rxn.gapfilling_type = 'universal'
+        for met in self.model.metabolites:
+            if exchange_reactions:
+                rxn = self.universal.add_boundary(
+                    met, type='exchange_smiley', lb=-1000, ub=0,
+                    reaction_id='SMILEY_EX_{}'.format(met.id))
```

Comment:
Since this is a generic gap fill algorithm, does it make sense for exchange and demand reactions to be prefixed with "SMILEY"?